test/Test/Toml/PrefixTree/Property.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Toml.PrefixTree.Property
+       ( propertyTests
+       ) where
+
+import Control.Monad (forM)
+
+import Hedgehog (MonadGen, Property, forAll, property, (===))
+import Test.Tasty (TestTree)
+import Test.Tasty.Hedgehog (testProperty)
+
+import Toml.PrefixTree (pattern (:||), Key (..), Piece (..), PrefixMap, PrefixTree (..))
+
+import qualified Data.HashMap.Strict as HashMap
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Toml.PrefixTree as Prefix
+
+propertyTests :: [TestTree]
+propertyTests = [insertLookup]
+
+insertLookup :: TestTree
+insertLookup = testProperty "lookup k (insert k v m) == Just v" prop_insertLookup
+
+----------------------------------------------------------------------------
+-- Common generators
+----------------------------------------------------------------------------
+
+type V = Int
+
+genVal :: MonadGen m => m V
+genVal = Gen.int (Range.constant 0 256)
+
+genPiece :: MonadGen m => m Piece
+genPiece = Piece <$> Gen.text (Range.constant 1 50) Gen.unicode
+
+genKey :: MonadGen m => m Key
+genKey = Key <$> Gen.nonEmpty (Range.constant 1 10) genPiece
+
+-- Generates key-value pair for PrefixMap
+genEntry :: MonadGen m => m (Piece, Key)
+genEntry = do
+    key@(piece :|| _) <- genKey
+    pure (piece, key)
+
+genPrefixMap :: MonadGen m => m (PrefixMap V)
+genPrefixMap = do
+    entries <- Gen.list (Range.linear 0 10) genEntry
+    kvps    <- forM entries $ \(piece, key) -> do
+        tree <- genPrefixTree key
+        pure (piece, tree)
+
+    pure $ HashMap.fromList kvps
+
+genPrefixTree :: forall m . MonadGen m => Key -> m (PrefixTree V)
+genPrefixTree key = Gen.recursive
+    -- list picker generator combinator
+    Gen.choice
+    -- non-recursive generators
+    [ Leaf key <$> genVal ]
+    -- recursive generators
+    [ genPrefixMap >>= genBranch ]
+  where
+    genBranch :: PrefixMap V -> m (PrefixTree V)
+    genBranch prefMap = do
+        prefVal <- Gen.maybe genVal
+        pure $ Branch key prefVal prefMap
+
+----------------------------------------------------------------------------
+-- InsertLookup
+----------------------------------------------------------------------------
+
+prop_insertLookup :: Property
+prop_insertLookup =  property $ do
+    prefMap <- forAll genPrefixMap
+    key     <- forAll genKey
+    val     <- forAll genVal
+
+    Prefix.lookup key (Prefix.insert key val prefMap) === Just val
+
+    -- DEBUG: ensures that trees of depth at least 5 are generated
+    -- assert $ depth prefMap < 5
+
```

Comment:
Another property I can think of is
```
lookup a (insert a y $ insert a x t) === Just y
```
to check that it uses the latest value. How do you think?

Reply:
@vrom911 That's an excellent property! I can add it as well.